modules/images/webp-uploads/load.php

```diff
@@ -11,42 +11,106 @@
  */
 function webp_uploads_create_sources_property( array $metadata, $attachment_id ) {
-	// Make sure we have some sizes to work with, otherwise avoid any work.
-	if ( empty( $metadata['sizes'] ) || ! is_array( $metadata['sizes'] ) ) {
-		return $metadata;
-	}
 	// This should take place only on the JPEG image.
 	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
-
 	// Not a supported mime type to create the sources property.
 	$mime_type = get_post_mime_type( $attachment_id );
-	if ( ! isset( $valid_mime_transforms[ $mime_type ] ) ) {
+	if ( ! isset( $valid_mime_transforms[ $mime_type ] ) || ! is_array( $valid_mime_transforms[ $mime_type ] ) ) {
 		return $metadata;
 	}
 
 	$file = get_attached_file( $attachment_id, true );
-
 	// File does not exist.
 	if ( ! file_exists( $file ) ) {
 		return $metadata;
 	}
+	// Make sure the top level `sources` key is a valid array.
+	if ( ! isset( $metadata['sources'] ) || ! is_array( $metadata['sources'] ) ) {
+		$metadata['sources'] = array();
+	}
+
+	if ( empty( $metadata['sources'][ $mime_type ] ) ) {
+		$metadata['sources'][ $mime_type ] = array(
+			'file'     => wp_basename( $file ),
+			'filesize' => filesize( $file ),
+		);
+		wp_update_attachment_metadata( $attachment_id, $metadata );
+	}
+
+	$allowed_mimes      = array_flip( wp_get_mime_types() );
+	$original_directory = pathinfo( $file, PATHINFO_DIRNAME );
+	$original_extension = pathinfo( $file, PATHINFO_EXTENSION );
+	// Create the sources for the full sized image.
+	foreach ( $valid_mime_transforms[ $mime_type ] as $targeted_mime ) {
+		// If this property exists no need to create the image again.
+		if ( ! empty( $metadata['sources'][ $targeted_mime ] ) ) {
+			continue;
+		}
+
+		if ( ! isset( $allowed_mimes[ $targeted_mime ] ) || ! is_string( $allowed_mimes[ $targeted_mime ] ) ) {
+			continue;
+		}
+
+		if ( ! wp_image_editor_supports( array( 'mime_type' => $targeted_mime ) ) ) {
+			continue;
+		}
+
+		$editor = wp_get_image_editor( $file );
+
+		if ( is_wp_error( $editor ) ) {
+			continue;
+		}
+
+		if ( ! isset( $metadata['width'], $metadata['height'] ) ) {
+			continue;
+		}
+
+		$editor->resize( $metadata['width'], $metadata['height'], false );
+		$extension = explode( '|', $allowed_mimes[ $targeted_mime ] );
+		$extension = reset( $extension );
+
+		if ( empty( $extension ) || empty( $original_extension ) ) {
+			continue;
+		}
+
+		$filename         = wp_basename( $file, ".{$original_extension}" );
+		$target_file_name = trailingslashit( $original_directory ) . "{$filename}.{$extension}";
+		$image            = $editor->save( $target_file_name, $targeted_mime );
+
+		if ( is_wp_error( $image ) ) {
+			continue;
+		}
+
+		$metadata['sources'][ $targeted_mime ] = array(
+			'file'     => $image['file'],
+			'filesize' => isset( $image['path'] ) ? filesize( $image['path'] ) : 0,
+		);
+
+		wp_update_attachment_metadata( $attachment_id, $metadata );
+	}
```

Comment:
There is _a lot_ of duplicate code in this loop which is very similar to the same `foreach ( $formats ...)` loop further below for generating the additional formats for the sub-sizes. We should simplify this and reuse some common functionality. Specifically, we should rework the `webp_uploads_generate_image_size()` function so that it can be reused for both scenarios, both generating the full image as well as the sub-sized images.

I suggest the following:
* Introduce a new `webp_uploads_generate_additional_image_source( $attachment_id, $src_file, $dst_file, $mime, $size_data )`, where:
    * `$attachment_id` should be the attachment ID. That may not even be used by the function, but it should be specified for consistency with other functions and most importantly, to have that context available e.g. for potential filters introduced in the function later.
    * `$src_file` should be the full file name (including path) to the source file to generate the image from (i.e. the original image file). This should still be a parameter though to allow for flexibility and to control this from outside the function, which e.g. will also help with tests.
    * `$dst_file` should be the full file name (including path) where to save the generated file.
    * `$mime` should be the MIME type to generate the new image.
    * `$size_data` should be an associative array with `width`, `height` (both int), and `crop` (bool), as typically specified for WordPress image sizes).
* This function should mostly take over code that is currently present in `webp_uploads_generate_image_size()`, and it should return the same data, an associative array with `file` and `filesize` on success, or a `WP_Error` on failure.
* The function should receive an `@access private` annotation, to indicate that it shouldn't be used by external developers.
* The `webp_uploads_generate_image_size()` should be simplified to use the above new function, so it will become a very simple function that just looks up the correct `$src_file`, `$dst_file`, and `$size_data` and then passes those to the other new function.
* The above `foreach` loop` should then also be simplified so that it just calls the above new function as well. This will result in code like the following:
```suggestion
	foreach ( $valid_mime_transforms[ $mime_type ] as $targeted_mime ) {
		// If this property exists no need to create the image again.
		if ( ! empty( $metadata['sources'][ $targeted_mime ] ) ) {
			continue;
		}

		if ( ! isset( $allowed_mimes[ $targeted_mime ] ) || ! is_string( $allowed_mimes[ $targeted_mime ] ) ) {
			continue;
		}

		if ( ! isset( $metadata['width'], $metadata['height'] ) ) {
			continue;
		}

		$extension        = explode( '|', $allowed_mimes[ $targeted_mime ] );
		$extension        = $extension[0];
		$filename         = wp_basename( $file, ".{$original_extension}" );
		$target_file_name = trailingslashit( $original_directory ) . "{$filename}.{$extension}";

		$size_data = array(
			'width'  => $metadata['width'],
			'height' => $metadata['width'],
			'crop'   => false,
		);

		$source = webp_uploads_generate_additional_image_source( $attachment_id, $file, $target_file_name, $targeted_mime, $size_data );
		if ( is_wp_error( $source ) ) {
			continue;
		}

		$metadata['sources'][ $targeted_mime ] = $source;
		wp_update_attachment_metadata( $attachment_id, $metadata );
	}
```

With the above, we will drastically reduce the duplicate code here and make it easier to maintain.

Reply:
Agree with that, originally this was a separate filter only to handle the `full size` but in order to keep things in a single place was merged into a single method.

---- 

This is the rationale behind not creating a new function or why not introducing additional parameters into the existing function.

The current logic on the method handles images sizes, and because the `full size` is a virtual size we can as you pointed out separate any depending logic from the sizes as parameters.

- The main problem with that was the fact that makes things more complex due number of parameters.
- The current logic to handle the sizes removes the suffix `scaled` or `rotated` if present, we can add and if block here to prevent this from happening where the image size is `full`. This won't be required anymore after #204 due the `scaled` was coming because we were using the `scaled` version in some cases, so that condition and replacement seem unnecessary after that PR is merged.

Possible problems when using the same function with additional parameters/conditionals inside: 

- Complexity increases due to the number of parameters
- Knowledge of the outside world by working differently in some cases.
- Problem when "full" size is defined by the user.

Agree that we can go deeper in breaking it down into smaller functions, but due we are using a global namespace this seems overkill, Lastly, I would argue that having duplicate code is better than having the wrong abstraction in place.

Happy to go down the route for more abstract modules but from my perspective it seems overkill for an additional image size, which effectively is not a size so that's why the abstraction does not fit.

Reply:
@mitogh I'm not sure I understand some of your points. There may be a misunderstanding related to my original comment.

> originally this was a separate filter

I'm not talking about a separate filter, the way you have it all in one function makes sense. I'm just saying we should introduce another low-level function to remove duplicated logic in the two different `foreach` loops which is similar for creating an image based on another image. That part is the same no matter if it's for the original image or for a sub-sized image.

> The current logic on the method handles images sizes

I'm not suggesting to change the parameters of that existing function, I'm suggesting to introduce a new _low-level_ function that can be used in both relevant plaes.

> The current logic to handle the sizes removes the suffix `scaled` or `rotated` if present, we can add and if block here to prevent this from happening where the image size is `full`.

With my suggestion from above nothing would change here. That logic would be kept in the `webp_uploads_generate_image_size()` function just like it is today. Again, I'm talking about a new function, not expanding the behavior of the existing `webp_uploads_generate_image_size()` function.

> Possible problems when using the same function with additional parameters/conditionals inside:

See above, I'm not saying we should use the same function with additional parameters.

> Agree that we can go deeper in breaking it down into smaller functions, but due we are using a global namespace this seems overkill, Lastly, I would argue that having duplicate code is better than having the wrong abstraction in place.

What about my suggested abstraction is wrong in your opinion, can you clarify? My proposed new function `webp_uploads_generate_additional_image_source()` has nothing to do with sub-sized images or the original image, it is 100% agnostic of that. It has a single clear purpose: Create an image based on a given source file in a given MIME type for given size arguments. It does nothing else than that, and it won't have any if clauses that would be different depending on whether it's a sub-size or the original image (or the `full` image). All the dynamic parts relating to that would still be handled outside of that function, because they are different, I agree with you. The common behavior that should be put into its own function to remove duplicate code is the context-unaware creation of a new image based on whatever parameters you pass into it.

Reply:
> I'm not suggesting to change the parameters of that existing function, I'm suggesting to introduce a new low-level function that can be used in both relevant places.

Mostly wanted to avoid adding new global functions but I do see the benefit in this approach to simplify and cleanup duplication, so let me go with this route instead.

> With my suggestion from above nothing would change here. That logic would be kept in the webp_uploads_generate_image_size() function just like it is today. Again, I'm talking about a new function, not expanding the behavior of the existing webp_uploads_generate_image_size() function.

:+1: 

> What about my suggested abstraction is wrong in your opinion, can you clarify? My proposed new function webp_uploads_generate_additional_image_source() has nothing to do with sub-sized images or the original image, it is 100% agnostic of that.

I think this one is on me as I confused the new function `webp_uploads_generate_additional_image_source` with `webp_uploads_generate_image_size` instead so I agree that this function makes more sense to keep it private. One thing to add is besides marking the function as private we also should follow the convention to prefix such functions with a dash like `_webp_uploads_generate_image_size` instead like: `_wp_admin_bar_init` or `_get_cron_lock`

Reply:
Thanks for clarifying, I had the impression this was a misunderstanding :)

> One thing to add is besides marking the function as private we also should follow the convention to prefix such functions with a dash

Note that this convention is not standardized in WP core. I have had multiple WP core committers tell me over the years that we shouldn't use the underscore prefix anymore, which makes sense to me. The `@private` is sufficient as an indicator, the underscore prefix doesn't _really_ mean anything. So I'd advise to leave the function name without underscore prefix.

Reply:
Thanks, just applied this feedback let me know in case you have any additional feedback on this one.